File: templates/api/springboot-api-java/src/main/java/com/prodstarter/Application.java
package com.prodstarter;

import io.micrometer.core.instrument.MeterRegistry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.ConfigurationPropertiesScan;
import org.springframework.context.ApplicationListener;
import org.springframework.context.event.ContextClosedEvent;
import org.springframework.context.annotation.Bean;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;
import org.springframework.web.filter.CorsFilter;
import org.springframework.web.filter.CommonsRequestLoggingFilter;

import java.time.Duration;
import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * ProdStarter — Spring Boot Application entrypoint (Java)
 *
 * Production-ready Spring Boot application main file with:
 *  - Micrometer MeterRegistry customization (common tags)
 *  - Request logging filter (CommonsRequestLoggingFilter)
 *  - CORS filter wired from property `app.cors.allowed-origins`
 *  - Graceful shutdown helper (ContextClosedEvent listener)
 *  - Global exception handler to avoid leaking internals
 *
 * Customize TODO blocks to add DB, security, tracing, or other integrations.
 */
@SpringBootApplication
@ConfigurationPropertiesScan
public class Application {

    private static final Logger log = LoggerFactory.getLogger(Application.class);

    // Managed executor for short-lived background tasks — prefer Spring TaskExecutor in prod
    private final ExecutorService backgroundExecutor = Executors.newCachedThreadPool();

    public static void main(String[] args) {
        // Default profile: prefer `production` for released templates; dev can override to `dev`
        String active = System.getProperty("spring.profiles.active");
        if (active == null || active.isBlank()) {
            System.setProperty("spring.profiles.default", "production");
        }

        SpringApplication.run(Application.class, args);
    }

    /**
     * Register common tags on MeterRegistry for consistent metrics labeling.
     */
    @Bean
    public MeterRegistry configureMeters(MeterRegistry registry, Environment env) {
        String appName = env.getProperty("spring.application.name", "prodstarter");
        String envName = env.getProperty("spring.profiles.active",
                env.getProperty("spring.profiles.default", "default"));

        registry.config().commonTags("application", appName, "environment", envName);
        return registry;
    }

    /**
     * Simple request logging filter for incoming HTTP requests. Disable payload logging in production
     * unless needed; it can be noisy and leak sensitive data.
     */
    @Bean
    public CommonsRequestLoggingFilter requestLoggingFilter(Environment env) {
        CommonsRequestLoggingFilter filter = new CommonsRequestLoggingFilter();
        filter.setIncludeClientInfo(true);
        filter.setIncludeQueryString(true);
        // set to false in prod; enable in debug when diagnosing issues
        boolean includePayload = Boolean.parseBoolean(env.getProperty("app.logging.include-payload", "false"));
        filter.setIncludePayload(includePayload);
        filter.setMaxPayloadLength(Integer.parseInt(env.getProperty("app.logging.max-payload-length", "1024")));
        filter.setIncludeHeaders(false);
        return filter;
    }

    /**
     * CORS filter configurable via property `app.cors.allowed-origins` (comma separated). In production you
     * should set explicit origins; default is permissive for development only.
     */
    @Bean
    public CorsFilter corsFilter(Environment env) {
        String origins = env.getProperty("app.cors.allowed-origins", "");
        CorsConfiguration config = new CorsConfiguration();
        if (origins == null || origins.isBlank()) {
            // permissive default for developer experience; override in production profile
            config.setAllowedOriginPatterns(Collections.singletonList("*"));
        } else {
            String[] parts = origins.split(",");
            Arrays.stream(parts).map(String::trim).filter(s -> !s.isEmpty()).forEach(config::addAllowedOrigin);
        }
        config.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"));
        config.setAllowedHeaders(Collections.singletonList("*"));
        config.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", config);
        return new CorsFilter(source);
    }

    /**
     * Graceful shutdown: listen for ContextClosedEvent and drain backgroundExecutor before JVM exits.
     * For Kubernetes, also enable `server.shutdown=graceful` in application properties to let the embedded
     * server drain connections properly.
     */
    @Bean
    public ApplicationListener<ContextClosedEvent> gracefulShutdownListener() {
        return event -> {
            log.info("ContextClosedEvent received — starting graceful shutdown of background executor");
            backgroundExecutor.shutdown();
            try {
                if (!backgroundExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
                    log.warn("Background executor didn't terminate in 30s — forcing shutdown");
                    backgroundExecutor.shutdownNow();
                }
            } catch (InterruptedException ex) {
                log.warn("Interrupted during graceful shutdown, forcing shutdown", ex);
                backgroundExecutor.shutdownNow();
                Thread.currentThread().interrupt();
            }
            log.info("Graceful shutdown finished");
        };
    }

    /**
     * Helper to submit short-lived background tasks via the managed executor. Prefer Spring's
     * TaskExecutor or @Async for production workloads; this helper is convenience for quick jobs.
     */
    public void submitBackgroundTask(Runnable task) {
        backgroundExecutor.submit(task);
    }

    /**
     * Global exception handler to centralize error responses and avoid leaking sensitive details.
     * For more advanced handling, create a dedicated @ControllerAdvice class per API group.
     */
    @ControllerAdvice
    public static class GlobalExceptionHandler {

        private final Logger log = LoggerFactory.getLogger(GlobalExceptionHandler.class);

        @ExceptionHandler(Exception.class)
        public ResponseEntity<Object> handleGeneric(Exception ex) {
            log.error("Unhandled exception caught", ex);
            // Do not expose internal errors in production payloads
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Collections.singletonMap("message", "An unexpected error occurred"));
        }

        // Add handlers for specific exceptions (e.g., validation, access denied) as needed.
    }

    // TODO: Add beans for DB connectivity, caching, tracing (OpenTelemetry), Sentry, security filters, etc.

}

File: templates/api/kotlin-springboot-api/src/main/kotlin/com/prodstarter/Application.kt
package com.prodstarter

import io.micrometer.core.instrument.MeterRegistry
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.context.properties.ConfigurationPropertiesScan
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.event.ContextClosedEvent
import org.springframework.context.event.EventListener
import org.springframework.core.env.Environment
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.reactive.CorsWebFilter
import org.springframework.web.cors.reactive.UrlBasedCorsConfigurationSource
import org.springframework.web.filter.CommonsRequestLoggingFilter
import org.springframework.web.server.ResponseStatusException
import java.time.Duration
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * ProdStarter — Spring Boot Application entrypoint
 *
 * Production-ready spring boot Kotlin application main file.
 * Features included/ready to wire:
 *  - Structured logging (SLF4J with Logback)
 *  - Micrometer metrics registration hook
 *  - Request logging filter (adjustable)
 *  - Global exception handler that avoids leaking internal details
 *  - Graceful shutdown helper (background executor drain)
 *  - CORS reactive filter configured from application properties
 *  - ConfigurationProperties scanning (for strongly-typed config classes)
 *  - Toggleable OpenAPI/Swagger (configure via application properties)
 *
 * Customize the TODO blocks to add DB, security, OAuth2/JWT, OpenTelemetry tracing, or any other integrations.
 */

@SpringBootApplication
@ConfigurationPropertiesScan
class Application {
    private val log = LoggerFactory.getLogger(Application::class.java)

    // A small thread pool to demonstrate graceful shutdown (for background tasks)
    private val backgroundExecutor: ExecutorService = Executors.newCachedThreadPool()

    /**
     * Example bean to register Micrometer meter customizations or tags for all metrics.
     * The MeterRegistry will be auto-configured by Spring Boot if Micrometer dependency is present.
     */
    @Bean
    fun configureMeters(registry: MeterRegistry, env: Environment): MeterRegistry {
        // Example: add common tags to all meters
        registry.config().commonTags("application", env.getProperty("spring.application.name", "prodstarter"),
            "environment", env.getProperty("spring.profiles.active", env.getProperty("spring.profiles.default", "default")))
        return registry
    }

    /**
     * Fine-grained request logging for debugging and audit. Disable or tune in production if noisy.
     * Uses CommonsRequestLoggingFilter (Servlet-based). For WebFlux, adapt to a reactive logging filter.
     */
    @Bean
    fun requestLoggingFilter(): CommonsRequestLoggingFilter {
        val filter = CommonsRequestLoggingFilter()
        filter.setIncludeClientInfo(true)
        filter.setIncludeQueryString(true)
        filter.setIncludePayload(false) // set true with payload limit for debug only
        filter.setMaxPayloadLength(1024)
        filter.setIncludeHeaders(false)
        return filter
    }

    /**
     * CORS configuration. In production, configure allowed origins via property `app.cors.allowed-origins`.
     */
    @Bean
    fun corsFilter(env: Environment): CorsWebFilter {
        val corsConfig = CorsConfiguration()
        val origins = env.getProperty("app.cors.allowed-origins")
        if (!origins.isNullOrBlank()) {
            origins.split(',').map { it.trim() }.forEach { corsConfig.addAllowedOrigin(it) }
        } else {
            // permissive default for local development — override in prod
            corsConfig.addAllowedOriginPattern("*")
        }
        corsConfig.allowedMethods = listOf("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
        corsConfig.allowedHeaders = listOf("*")
        corsConfig.allowCredentials = true

        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", corsConfig)
        return CorsWebFilter(source)
    }

    /**
     * Example: listen for application context close to perform graceful shutdown tasks.
     * For containers orchestrators (Kubernetes), enable `server.shutdown=graceful` in properties
     * and tune `spring.lifecycle.timeout-per-shutdown-phase` for longer drains.
     */
    @EventListener(ContextClosedEvent::class)
    fun onContextClosed(event: ContextClosedEvent) {
        log.info("ContextClosedEvent received — starting graceful shutdown of background executor")
        backgroundExecutor.shutdown()
        try {
            if (!backgroundExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
                log.warn("Background executor didn't terminate in 30s — forcing shutdown")
                backgroundExecutor.shutdownNow()
            }
        } catch (ex: InterruptedException) {
            log.warn("Interrupted during graceful shutdown, forcing shutdown", ex)
            backgroundExecutor.shutdownNow()
            Thread.currentThread().interrupt()
        }
        log.info("Graceful shutdown finished")
    }

    /**
     * Helper to submit short-lived background tasks via the managed executor.
     * Use application-scoped task schedulers or Spring @Async (with Executor config) for production.
     */
    fun submitBackgroundTask(runnable: Runnable) {
        backgroundExecutor.submit(runnable)
    }

    /**
     * Global exception handler to centralize error responses and avoid leaking sensitive details.
     * For more advanced handling, create separate @ControllerAdvice classes per api-group.
     */
    @ControllerAdvice
    class GlobalExceptionHandler {
        private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

        @ExceptionHandler(Exception::class)
        fun handleAny(ex: Exception): ResponseEntity<Map<String, Any>> {
            // Log full stack trace with structured logger
            log.error("Unhandled exception caught", ex)

            val body = mapOf(
                "error" to "internal_server_error",
                "message" to "An unexpected error occurred."
            )
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(body)
        }

        @ExceptionHandler(ResponseStatusException::class)
        fun handleResponseStatus(ex: ResponseStatusException): ResponseEntity<Map<String, Any>> {
            val body = mapOf(
                "error" to ex.status.reasonPhrase?.lowercase()?.replace(" ", "_") ?: "error",
                "message" to (ex.reason ?: ex.message ?: "")
            )
            return ResponseEntity.status(ex.status).body(body)
        }
    }
}

fun main(args: Array<String>) {
    // Set a sensible default profile if none active — prefer 'prod' for releases
    val activeProfiles = System.getProperty("spring.profiles.active") ?: System.getenv("SPRING_PROFILES_ACTIVE")
    if (activeProfiles.isNullOrBlank()) {
        // default to 'production' for template release; developers can use 'dev' locally
        System.setProperty("spring.profiles.default", "production")
    }

    // Start the application
    runApplication<Application>(*args) {
        // Additional programmatic customizations can be set here.
        addInitializers()
    }
}
